src/coprocessor/util.rs

```diff
@@ -17,33 +17,89 @@ use tipb::schema::ColumnInfo;
     }
-    let mut i = nk.len() - 1;
+    let mut i = key.len() - 1;
+
+    // Add 1 to the last byte that is not 255, and set it's following bytes to 0.
     loop {
-        if nk[i] == 255 {
-            nk[i] = 0;
+        if key[i] == 255 {
+            key[i] = 0;
         } else {
-            nk[i] += 1;
-            return nk;
+            key[i] += 1;
+            return;
         }
         if i == 0 {
-            nk = key.to_vec();
-            nk.push(0);
-            return nk;
+            // All bytes are 255. Append a 0 to the key.
+            for byte in key.iter_mut() {
+                *byte = 255;
+            }
+            key.push(0);
+            return;
         }
         i -= 1;
     }
 }
 
+/// Check if `key`'s prefix next equals to `next`
+pub fn is_prefix_next(key: &[u8], next: &[u8]) -> bool {
+    let len = key.len();
+    let next_len = next.len();
+
+    if len == next_len {
+        // Find the last non-255 byte
+        let mut carry_pos = len;
+        loop {
+            if carry_pos == 0 {
+                // All bytes of `key` are 255. `next` couldn't be `key`'s prefix_next since their
+                // lengths are equal.
+                return false;
+            }
+
+            carry_pos -= 1;
+            if key[carry_pos] != 255 {
+                break;
+            }
+        }
+
+        // Now `carry_pos` is the index of the last byte that is not 255. For example:
+        //   key: [1, 2, 3, 255, 255, 255]
+        //               ^ carry_pos == 2
+
+        // So they are equal when:
+        // * `key`'s value at `carry_pos` is that of `next` - 1 and
+        // * `next`'s part after carry_pos is all 0
+        // * `key` and `next`'s parts before `carry_pos` are equal.
+        // For example:
+        //   key:  [1, 2, 3, 255, 255, 255]
+        //   next: [1, 2, 4,   0,   0,   0]
+        //                ^ carry_pos == 2
+        // The part before `carry_pos` is all [1, 2],
+        // the bytes at `carry_pos` differs by 1 (4 == 3 + 1), and
+        // the remaining bytes of next ([0, 0, 0]) is all 0.
+        // so [1, 2, 4, 0, 0, 0] is prefix_next of [1, 2, 3, 255, 255, 255]
+        key[carry_pos] + 1 == next[carry_pos]
+            && next[carry_pos + 1..].iter().all(|byte| *byte == 0)
+            && key[..carry_pos] == next[..carry_pos]
+    } else if len + 1 == next_len {
+        // `next` must has one more 0 than `key`, and the first `len` bytes must be all 255.
+        // The case that `len == 0` is also covered here.
+        *next.last().unwrap() == 0
+            && key.iter().all(|byte| *byte == 255)
+            && next.iter().take(len).all(|byte| *byte == 255)
+    } else {
+        // Length not match.
+        false
+    }
+}
+
 /// `is_point` checks if the key range represents a point.
 pub fn is_point(range: &coppb::KeyRange) -> bool {
-    range.get_end() == &*prefix_next(range.get_start())
+    is_prefix_next(range.get_start(), range.get_end())
```

Comment:
Please migrate this change (including newly added tests) into master as well. @MyonKeminta

Reply:
Got it